sql: add summarized query formats for select, insert, and update statements

lindseyjin opened the pull request:
Resolves #27021

Previously, statements on the statements page hid too much information.
There were complaints that it was difficult to disambiguate between
statements without having to view the full query on the tooltips.

To address this, this commit adds in new formatting functions for
creating summarized queries. Currently, this only applies to SELECT,
INSERT, and UPDATE statement types. We have also added a new FmtFlag,
FmtSummary, a new field, StmtSummary, in the Statement struct, and a
new field, QuerySummary, in app stats. A later commit will implement
these changes in the frontend.

Release note (sql change): Add new formatting functions to create
summarized queries for SELECT, INSERT, and UPDATE statements. Also
add new metadata fields, which will later be used to pass this
information to the front-end Statements page.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/71012)
<!-- Reviewable:end -->

lindseyjin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/ui/workspaces/cluster-ui/src/statementsTable/statementsTable.tsx, line 244 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlBAJfI1ZX9kvKHt-Ix:-MlBAJfI1ZX9kvKHt-Iy:b77cbx) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5eddd61e59c5ae18bce52c5a34a0da557dbe040/pkg/ui/workspaces/cluster-ui/src/statementsTable/statementsTable.tsx#L244)):*
> ```XML
>   original: string,
> ) {
>   switch (summary.statement) {
> ```

I wanted to get something up for early review, just to make sure I'm approaching this correctly! I can add more cases for the other formats we want to implement later (insert, update), but wondering if I should do that in the same PR, or a separate one, or in separate commits? 
___
*[pkg/ui/workspaces/cluster-ui/src/util/sql/summarize.ts, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlBAhqe0D39PeOGGz70:-MlBAhqe0D39PeOGGz71:b-tjb97t) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5eddd61e59c5ae18bce52c5a34a0da557dbe040/pkg/ui/workspaces/cluster-ui/src/util/sql/summarize.ts#L66)):*
> ```XML
>   };
> 
>   const sqlKeywords = new Set(["ON", "WHERE", "ORDER BY"]);
> ```

Is there an easier, more-effective way to separate SQL clauses? From a quick glance, I think my current solution is able to catch most cases I see on the Statements page, although definitely don't think I'm covering every edge case right now. 


<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @lindseyjin, and @maryliag)

___
*[pkg/ui/workspaces/cluster-ui/src/statementsTable/statementsTable.tsx, line 243 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlBFpAQAobAq-BssF-U:-MlBFpAQAobAq-BssF-V:b-39sb0u) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5eddd61e59c5ae18bce52c5a34a0da557dbe040/pkg/ui/workspaces/cluster-ui/src/statementsTable/statementsTable.tsx#L243)):*
> ```XML
>   summary: StatementSummary,
>   original: string,
> ) {
> ```

nit: let's annotated this with return types :D 
___
*[pkg/ui/workspaces/cluster-ui/src/util/sql/summarize.ts, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlBAhqe0D39PeOGGz70:-MlBGeCv1184owLpDJMF:b-1bdyou) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5eddd61e59c5ae18bce52c5a34a0da557dbe040/pkg/ui/workspaces/cluster-ui/src/util/sql/summarize.ts#L66)):*
<details><summary><i>Previously, lindseyjin (Lindsey Jin) wrote…</i></summary><blockquote>

Is there an easier, more-effective way to separate SQL clauses? From a quick glance, I think my current solution is able to catch most cases I see on the Statements page, although definitely don't think I'm covering every edge case right now. 
</blockquote></details>

Hmm the only way to 100% correctly parse our SQL statements is to reuse the SQL parser used in the DB itself. This makes me think that, unless we want to reimplement the parsing logic in TS, maybe the best way for us to surface this information in the backend (where we actually have the parsed AST of a statement), and bundle that in the statement metadata. 

Our statement syntax also is constantly evolving, e.g. `AS OF SYSTEM TIME` clause is not really standard in Postgres and it is specific to cockroach db. 

cc @maryliag  what do you think? 


<!-- Sent from Reviewable.io -->

maryliag reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @lindseyjin, and @maryliag)

___
*[pkg/ui/workspaces/cluster-ui/src/statementsTable/statementsTable.tsx, line 244 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlBAJfI1ZX9kvKHt-Ix:-MlBHgna9upExvRijXhD:bbqo13z) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5eddd61e59c5ae18bce52c5a34a0da557dbe040/pkg/ui/workspaces/cluster-ui/src/statementsTable/statementsTable.tsx#L244)):*
<details><summary><i>Previously, lindseyjin (Lindsey Jin) wrote…</i></summary><blockquote>

I wanted to get something up for early review, just to make sure I'm approaching this correctly! I can add more cases for the other formats we want to implement later (insert, update), but wondering if I should do that in the same PR, or a separate one, or in separate commits? 
</blockquote></details>

you can do on the same PR and on the same commit
___
*[pkg/ui/workspaces/cluster-ui/src/util/sql/summarize.ts, line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlBHsKTEfEd0sw_xzs2:-MlBHsKTEfEd0sw_xzs3:b-6nvjz2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5eddd61e59c5ae18bce52c5a34a0da557dbe040/pkg/ui/workspaces/cluster-ui/src/util/sql/summarize.ts#L60)):*
> ```XML
> 
> // detailedSummarize takes a string SQL statement and produces a
> // structured summary of the query, with columns and join/nested table info
> ```

nit: all comment should end with a period
___
*[pkg/ui/workspaces/cluster-ui/src/util/sql/summarize.ts, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlBAhqe0D39PeOGGz70:-MlBIXlY7xh4posepu_v:b-9rn2i9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5eddd61e59c5ae18bce52c5a34a0da557dbe040/pkg/ui/workspaces/cluster-ui/src/util/sql/summarize.ts#L66)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Hmm the only way to 100% correctly parse our SQL statements is to reuse the SQL parser used in the DB itself. This makes me think that, unless we want to reimplement the parsing logic in TS, maybe the best way for us to surface this information in the backend (where we actually have the parsed AST of a statement), and bundle that in the statement metadata. 

Our statement syntax also is constantly evolving, e.g. `AS OF SYSTEM TIME` clause is not really standard in Postgres and it is specific to cockroach db. 

cc @maryliag  what do you think? 
</blockquote></details>

We don't want to recreate all the logic on the frontend, so I agree we can try use something from the backend and send a few new parameters, meaning keep the statement value as is, so it can be used on places we already use, and add a new metadata field that have the statement separated and that you can here later on
___
*[pkg/ui/workspaces/cluster-ui/src/util/sql/summarize.ts, line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlBJ3TDBKWPfiOvYTuo:-MlBJ3TDBKWPfiOvYTup:b-8so11) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5eddd61e59c5ae18bce52c5a34a0da557dbe040/pkg/ui/workspaces/cluster-ui/src/util/sql/summarize.ts#L87)):*
> ```XML
>       statementMatch.groups.table
>     ) {
>       // get columns from regex
> ```

nit: period
___
*[pkg/ui/workspaces/cluster-ui/src/util/sql/summarize.ts, line 98 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlBJ4ti6vDMlq1V9S6Z:-MlBJ4tjA7Phejm41VOz:b-8so11) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5eddd61e59c5ae18bce52c5a34a0da557dbe040/pkg/ui/workspaces/cluster-ui/src/util/sql/summarize.ts#L98)):*
> ```XML
>       }
> 
>       // get tables from regex
> ```

nit: period


<!-- Sent from Reviewable.io -->

lindseyjin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @lindseyjin, and @maryliag)

___
*[pkg/ui/workspaces/cluster-ui/src/util/sql/summarize.ts, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlBAhqe0D39PeOGGz70:-MlBJn_qDCXrY0Bxyira:bxargq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5eddd61e59c5ae18bce52c5a34a0da557dbe040/pkg/ui/workspaces/cluster-ui/src/util/sql/summarize.ts#L66)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

We don't want to recreate all the logic on the frontend, so I agree we can try use something from the backend and send a few new parameters, meaning keep the statement value as is, so it can be used on places we already use, and add a new metadata field that have the statement separated and that you can here later on
</blockquote></details>

got it, I can look into that! btw, do you have any code pointers on hand for any files that might be relevant? f not, I can try to do some digging myself first


<!-- Sent from Reviewable.io -->

lindseyjin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng and @maryliag)

___
*[pkg/ui/workspaces/cluster-ui/src/statementsTable/statementsTable.tsx, line 244 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlBAJfI1ZX9kvKHt-Ix:-MlBK2sE9Ug-Gu1fMIUh:b3kwnt1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5eddd61e59c5ae18bce52c5a34a0da557dbe040/pkg/ui/workspaces/cluster-ui/src/statementsTable/statementsTable.tsx#L244)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

you can do on the same PR and on the same commit
</blockquote></details>

gotcha :+1:


<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @lindseyjin, and @maryliag)

___
*[pkg/ui/workspaces/cluster-ui/src/util/sql/summarize.ts, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlBAhqe0D39PeOGGz70:-MlBKBr44Uaalm2FczKD:bsdd98n) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5eddd61e59c5ae18bce52c5a34a0da557dbe040/pkg/ui/workspaces/cluster-ui/src/util/sql/summarize.ts#L66)):*
<details><summary><i>Previously, lindseyjin (Lindsey Jin) wrote…</i></summary><blockquote>

got it, I can look into that! btw, do you have any code pointers on hand for any files that might be relevant? f not, I can try to do some digging myself first
</blockquote></details>

I think one place we can start doing some digging is the [`sql.Statement`](https://github.com/cockroachdb/cockroach/blob/1944c522001ee57de7c419397db7993e14fc0e95/pkg/sql/statement.go#L19) struct, which is what we use to represent statements within SQL package. `sql.Statement` embeds [`parser.Statement`](https://github.com/cockroachdb/cockroach/blob/cb877947be5bbe14465537126425cf6b5e96888b/pkg/sql/parser/parse.go#L42) which in turn embeds [`tree.Statement`](https://github.com/cockroachdb/cockroach/blob/170de08ab1b6978b7768e79856a71283cd142d7f/pkg/sql/sem/tree/stmt.go#L89) interface.

`tree.Statement` interface is what we use to represent the statement AST. It contains `NodeFormatter` interface which walks through the AST tree and format statement into string based on formatting flags. 

Thomas did something like this for statement reduction back then in [this PR](https://github.com/cockroachdb/cockroach/pull/66359), i think if it's me I'd start digging there :P 


<!-- Sent from Reviewable.io -->

lindseyjin committed:
sql: add summarized query formats for select, insert, and update statements

Resolves #27021

Previously, statements on the statements page hid too much information.
There were complaints that it was difficult to disambiguate between
statements without having to view the full query on the tooltips.

To address this, this commit adds in new formatting functions for
creating summarized queries. Currently, this only applies to SELECT,
INSERT, and UPDATE statement types. We have also added a new FmtFlag,
FmtSummary, a new field, StmtSummary, in the Statement struct, and a
new field, QuerySummary, in app stats. A later commit will implement
these changes in the frontend.

Release note (sql change): Add new formatting functions to create
summarized queries for SELECT, INSERT, and UPDATE statements. Also
add new metadata fields, which will later be used to pass this
information to the front-end Statements page.

lindseyjin commented:
-

lindseyjin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @lindseyjin, and @maryliag)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlW9qS6BaMjcbJNEgEI:-MlW9qS6BaMjcbJNEgEJ:b-c8fdyk) (no related file):*
Trying to break my code down into smaller commits, so I'll be including metadata changes and frontend changes in a later commit. Also haven't implemented the JOIN simplifying logic yet, since it seemed a little complicated. Currently clarifying some details with Anne and Kevin, but I can add that in a future commit as well. 

---
___
*[pkg/sql/sem/tree/format.go, line 544 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlWA_cyC95SEGbXi_l1:-MlWA_cyC95SEGbXi_l2:b9615yf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0b2be1f0f613aa93155d1c856e53cc4d5b593e2/pkg/sql/sem/tree/format.go#L544)):*
> ```Go
> 
> // FormatNodeSummary recurses into a node for pretty-printing a summarized version.
> func (ctx *FmtCtx) FormatNodeSummary(n NodeFormatter) {
> ```

Hmm, I'm not sure if it would've been cleaner to pass down my FmtSummary flag and implement the summarized versions in each select/insert/update Format function instead of in my own functions? Although I didn't want to have too many code snippets everywhere, and there's also the issue with outer and inner SELECT clauses being formatted differently. 
___
*[pkg/sql/sem/tree/format_test.go, line 447 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlW9ZSk8gw3qAMM-pxc:-MlW9ZSk8gw3qAMM-pxd:b-70vfjl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0b2be1f0f613aa93155d1c856e53cc4d5b593e2/pkg/sql/sem/tree/format_test.go#L447)):*
> ```Go
> 		{`INSERT INTO vehicles (vehicles, rides) VALUES (1), (2), (3), (4), (5)`,
> 			`INSERT INTO vehicles (vehicles, rides)`},
> 		{`INSERT INTO vehicles (vehicles, rides) SELECT vehicle_ids, rides FROM riders`,
> ```

I wasn't able to test with periods in my strings, since I was getting an error message about the format not being implemented. Is there a way around this?
___
*[pkg/sql/sem/tree/select.go, line 93 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlWADG-D1Ww66zKH1g3:-MlWADG-D1Ww66zKH1g4:b8edckd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0b2be1f0f613aa93155d1c856e53cc4d5b593e2/pkg/sql/sem/tree/select.go#L93)):*
> ```Go
> // Format implements the NodeFormatter interface.
> func (node *SelectClause) Format(ctx *FmtCtx) {
> 	f := ctx.flags
> ```

Note: This is a special case for nested Select statements, where we don't wanna show columns at all. 


<!-- Sent from Reviewable.io -->

matthewtodd reviewed (changes requested):
> Reviewed 3 of 5 files at r5, 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @lindseyjin, @maryliag, and @matthewtodd)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlW9qS6BaMjcbJNEgEI:-MlkOMJH80I7W6zNrlgu:b-6ha7a7) (no related file):*
<details><summary><i>Previously, lindseyjin (Lindsey Jin) wrote…</i></summary><blockquote>

Trying to break my code down into smaller commits, so I'll be including metadata changes and frontend changes in a later commit. Also haven't implemented the JOIN simplifying logic yet, since it seemed a little complicated. Currently clarifying some details with Anne and Kevin, but I can add that in a future commit as well. 
</blockquote></details>

Yay for small, focused changes! :tada: 

---
___
*[pkg/sql/sem/tree/format.go, line 579 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlkPbA166J3A8KQEv-c:-MlkPbA166J3A8KQEv-d:b-h3fn6) ([raw file](https://github.com/cockroachdb/cockroach/blob/1601a9680681fd7288d3d4f3eec538adfa671aea/pkg/sql/sem/tree/format.go#L579)):*
> ```Go
> // AsShortStringWithFlags pretty prints a node to a string given specific flags; only
> // flags that don't require Annotations can be used.
> func AsShortStringWithFlags(n NodeFormatter, fl FmtFlags, opts ...FmtCtxOption) string {
> ```

Since `FmtSummary` is a new flag we're introducing here, I'm wondering if it would be clearer to fold this behavior into the existing `AsStringWithFlags`, so that callers don't have to both set the flag and call a different method. WDYT?


<!-- Sent from Reviewable.io -->

lindseyjin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @lindseyjin, @maryliag, and @matthewtodd)

___
*[pkg/sql/sem/tree/format.go, line 579 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlkPbA166J3A8KQEv-c:-MlpA2hFA1dl2kUCzl5E:b-mhxl2d) ([raw file](https://github.com/cockroachdb/cockroach/blob/1601a9680681fd7288d3d4f3eec538adfa671aea/pkg/sql/sem/tree/format.go#L579)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

Since `FmtSummary` is a new flag we're introducing here, I'm wondering if it would be clearer to fold this behavior into the existing `AsStringWithFlags`, so that callers don't have to both set the flag and call a different method. WDYT?
</blockquote></details>

Done! Thanks for the suggestion


<!-- Sent from Reviewable.io -->

matthewtodd reviewed (changes requested):
> Reviewed 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @lindseyjin, @maryliag, and @matthewtodd)

___
*[pkg/sql/sem/tree/format.go, line 490 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlpBrw_9vQBrr_xuxL1:-MlpBrw_9vQBrr_xuxL2:b-w46w1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cb22b4804ab56a8d35782ad2699b8a6103c7bac/pkg/sql/sem/tree/format.go#L490)):*
> ```Go
> // FormatLimitLength recurses into a node for pretty-printing, but limits the
> // number of characters to be printed.
> func (ctx *FmtCtx) FormatLimitLength(n NodeFormatter, maxLength int) {
> ```

One more thought, WDYT about making these `Format...` methods private?


<!-- Sent from Reviewable.io -->

lindseyjin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @lindseyjin, @maryliag, and @matthewtodd)

___
*[pkg/sql/sem/tree/format.go, line 490 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlpBrw_9vQBrr_xuxL1:-MlpPsmI2k1DBehHiR_z:b-jsolvv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cb22b4804ab56a8d35782ad2699b8a6103c7bac/pkg/sql/sem/tree/format.go#L490)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

One more thought, WDYT about making these `Format...` methods private?
</blockquote></details>

That's just a matter of making the functions lower case so they're not available outside the file, right? 

Done!


<!-- Sent from Reviewable.io -->

matthewtodd reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-:-MlpQ_r96WS87pEicrhi:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @Azhng, @lindseyjin, @maryliag, and @matthewtodd)

___
*[pkg/sql/sem/tree/format.go, line 490 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-MlpBrw_9vQBrr_xuxL1:-MlpQU8C2xxHpIyp_pTW:bb7867m) ([raw file](https://github.com/cockroachdb/cockroach/blob/7cb22b4804ab56a8d35782ad2699b8a6103c7bac/pkg/sql/sem/tree/format.go#L490)):*
<details><summary><i>Previously, lindseyjin (Lindsey Jin) wrote…</i></summary><blockquote>

That's just a matter of making the functions lower case so they're not available outside the file, right? 

Done!
</blockquote></details>

Yup!


<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @Azhng, @lindseyjin, @maryliag, and @matthewtodd)

___
*[pkg/sql/sem/tree/format_test.go, line 443 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-Mlq3xHy73JybSm5QJ4d:-Mlq3xHzCZFtA7KG7kvb:b-w2ptig) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f4ef205cd373a7d98bab3440b2ac271c0b658e8/pkg/sql/sem/tree/format_test.go#L443)):*
> ```Go
> 		// calls encodeSQLString with the right formatter.
> 		// See TestFormatExprs below for the test on DStrings.
> 		{`SELECT (SELECT count(*) FROM system.jobs) AS j, num_running, s.* FROM system.scheduled_jobs AS s WHERE next_run < current_timestamp() ORDER BY random() LIMIT 10 FOR UPDATE`,
> ```

It will really improve the readability a lot if you explicitly list out the field names, e.g.


``` golang
{
  stmt:           `SELECT ...`,
  expected:   `SELECT ....`,
}
```
___
*[pkg/sql/sem/tree/format_test.go, line 444 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-Mlq4O9S2efL6rsLWENg:-Mlq4O9S2efL6rsLWENh:b-1yuwlo) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f4ef205cd373a7d98bab3440b2ac271c0b658e8/pkg/sql/sem/tree/format_test.go#L444)):*
> ```Go
> 		// See TestFormatExprs below for the test on DStrings.
> 		{`SELECT (SELECT count(*) FROM system.jobs) AS j, num_running, s.* FROM system.scheduled_jobs AS s WHERE next_run < current_timestamp() ORDER BY random() LIMIT 10 FOR UPDATE`,
> 			`SELECT (SELECT FROM sy... FROM system.scheduled_jobs AS s`},
> ```

Hmm, I feel we might want to close the parenthesis for the subquery here, thoughts?  


<!-- Sent from Reviewable.io -->

lindseyjin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @lindseyjin, @maryliag, and @matthewtodd)

___
*[pkg/sql/sem/tree/format_test.go, line 443 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-Mlq3xHy73JybSm5QJ4d:-MlqR28d1ettxTaVXBfs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f4ef205cd373a7d98bab3440b2ac271c0b658e8/pkg/sql/sem/tree/format_test.go#L443)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

It will really improve the readability a lot if you explicitly list out the field names, e.g.


``` golang
{
  stmt:           `SELECT ...`,
  expected:   `SELECT ....`,
}
```
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/format_test.go, line 444 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-Mlq4O9S2efL6rsLWENg:-MlqPRMODlyDJF9v90PG:bb0s6sq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f4ef205cd373a7d98bab3440b2ac271c0b658e8/pkg/sql/sem/tree/format_test.go#L444)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Hmm, I feel we might want to close the parenthesis for the subquery here, thoughts?  
</blockquote></details>

Would require some refactoring, but I can add that in if you think it's necessary!

I'm also concerned about cases with a subquery, but also additional truncated text outside that query? eg) For the above example, it might be misleading to add a closing parenthesis

`SELECT (SELECT FROM sy...) FROM ...`

Because there's additional columns being selected after the subquery. Unless we also decide to add extra ellipses after the closing parenthesis, but that might be unnecessarily complicated.

`SELECT (SELECT FROM sy...)... FROM ...`

Thoughts?


<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @kevin-v-ngo, @knz, @lindseyjin, and @matthewtodd)

___
*[pkg/sql/sem/tree/format_test.go, line 444 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-Mlq4O9S2efL6rsLWENg:-MlqTPQHBPtuzzFo7OGf:b-fjh7r1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f4ef205cd373a7d98bab3440b2ac271c0b658e8/pkg/sql/sem/tree/format_test.go#L444)):*
<details><summary><i>Previously, lindseyjin (Lindsey Jin) wrote…</i></summary><blockquote>

Would require some refactoring, but I can add that in if you think it's necessary!

I'm also concerned about cases with a subquery, but also additional truncated text outside that query? eg) For the above example, it might be misleading to add a closing parenthesis

`SELECT (SELECT FROM sy...) FROM ...`

Because there's additional columns being selected after the subquery. Unless we also decide to add extra ellipses after the closing parenthesis, but that might be unnecessarily complicated.

`SELECT (SELECT FROM sy...)... FROM ...`

Thoughts?
</blockquote></details>

Hmm, personally, I found the summary string of the shape `SELECT (SELECT FROM sy...)... FROM ...` significantly more intuitive to read compared to the case of unbalanced parenthesis  (e.g. `SELECT (SELECT FROM sy... FROM system.scheduled_jobs AS s`). Defer to @kevin-v-ngo on whether if this really affect UX 

Hmm also, how come this is not being treated as the special case where we simply reduce it to `(SELECT)` ? 


<!-- Sent from Reviewable.io -->

maryliag reviewed (commented):
> Reviewed 3 of 5 files at r5, 1 of 3 files at r8, 1 of 1 files at r9, 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @kevin-v-ngo, @lindseyjin, @maryliag, and @matthewtodd)

___
*[pkg/sql/sem/tree/format_test.go, line 433 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-Mltp2IvDE45JjBpuy8b:-Mltp2IvDE45JjBpuy8c:b-qzqy1v) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6f1b913fc07d7994c298778d070f84ace0e7949/pkg/sql/sem/tree/format_test.go#L433)):*
> ```Go
> }
> 
> func TestFormatNodeSummary(t *testing.T) {
> ```

@kevin-v-ngo @annebirzin can you take a look into this test file just to validate the new formats or if there is any other case you have in mind that you would like to see the result.

@lindseyjin just confirming, you haven't dealt with the JOIN case yet, correct?


<!-- Sent from Reviewable.io -->

lindseyjin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Annebirzin, @Azhng, @kevin-v-ngo, @maryliag, and @matthewtodd)

___
*[pkg/sql/sem/tree/format_test.go, line 433 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-Mltp2IvDE45JjBpuy8b:-MltyaGE2ZPaNyv1Ru-y:btrao3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6f1b913fc07d7994c298778d070f84ace0e7949/pkg/sql/sem/tree/format_test.go#L433)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

@kevin-v-ngo @annebirzin can you take a look into this test file just to validate the new formats or if there is any other case you have in mind that you would like to see the result.

@lindseyjin just confirming, you haven't dealt with the JOIN case yet, correct?
</blockquote></details>

Oh, quick update for that: I talked to Kevin and Anne and they decided it was fine if we just go up to 30 characters starting from the FROM clause for tables, instead of implementing specific JOIN logic. That would be easier, and we’d likely capture the first table, the JOIN clause, ON clause, and maybe another table, before the text is truncated. 


<!-- Sent from Reviewable.io -->

maryliag reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-:-Mlu7TGI6Yud51OQKakh:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @Annebirzin, @Azhng, @kevin-v-ngo, @maryliag, and @matthewtodd)



<!-- Sent from Reviewable.io -->

lindseyjin reviewed (commented):
Hmm, I'm not too sure how it would render in the frontend, although I'd imagine that overly long statements would be cropped anyways instead of extending the table cell size. Can include screenshots in my next commit after I finish implementing that!

For now, I've updated the table character limit to be 15 instead of 30 characters. Is that better?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-:-MluS_dTE5R9tUyeWgmo:brkux8x)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @Annebirzin, @Azhng, @kevin-v-ngo, @maryliag, and @matthewtodd)



<!-- Sent from Reviewable.io -->

Annebirzin reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-:-MlubOv17IoR3kb5nUeB:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @Annebirzin, @Azhng, @kevin-v-ngo, @maryliag, and @matthewtodd)



<!-- Sent from Reviewable.io -->

lindseyjin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 3 stale) (waiting on @Annebirzin, @Azhng, @kevin-v-ngo, @maryliag, and @matthewtodd)

___
*[pkg/sql/sem/tree/format_test.go, line 444 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-Mlq4O9S2efL6rsLWENg:-Mm-s2qW-5FsftAwIKM0:b-cupy6z) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f4ef205cd373a7d98bab3440b2ac271c0b658e8/pkg/sql/sem/tree/format_test.go#L444)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Hmm, personally, I found the summary string of the shape `SELECT (SELECT FROM sy...)... FROM ...` significantly more intuitive to read compared to the case of unbalanced parenthesis  (e.g. `SELECT (SELECT FROM sy... FROM system.scheduled_jobs AS s`). Defer to @kevin-v-ngo on whether if this really affect UX 

Hmm also, how come this is not being treated as the special case where we simply reduce it to `(SELECT)` ? 

</blockquote></details>

Update to provide context: We're going with the first approach of closing the parentheses, and adding extra ellipses if necessary to signify additional text.

`SELECT (SELECT FROM sy...)... FROM ...`
___
*[pkg/sql/sem/tree/format_test.go, line 494 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-Mm0-gJF2KvRQIcV6s6s:-Mm0-gJF2KvRQIcV6s6t:bh01a5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/74088e382cdf3e0f709273317bff5b64a41d1bfe/pkg/sql/sem/tree/format_test.go#L494)):*
> ```Go
> 		{
> 			stmt:     `SELECT (SELECT job_id FROM (SELECT * FROM system.jobs) AS c) AS j, name, app_name FROM system.apps`,
> 			expected: `SELECT (SELECT FROM (S...)...)... FROM system.apps`,
> ```

Any thoughts on this format? @annebirzin @kevin-v-ngo 

Do we want to close all open parentheses and add ellipses if there's text afterwards?


<!-- Sent from Reviewable.io -->

lindseyjin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 3 stale) (waiting on @Annebirzin, @Azhng, @kevin-v-ngo, @maryliag, and @matthewtodd)

___
*[pkg/sql/sem/tree/format_test.go, line 462 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-Mm053et6wJ1Sc-wqDFf:-Mm053et6wJ1Sc-wqDFg:b-o5srzc) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba94984c27b66e8a355d16c64f24a8a9e5ceb9b9/pkg/sql/sem/tree/format_test.go#L462)):*
> ```Go
> 		{
> 			stmt:     `INSERT INTO vehicles VALUES ($1, $2, __more6__)`,
> 			expected: `INSERT INTO vehicles`,
> ```

I realized that insert statements without provided columns simplify down to just `INSERT INTO table`. Do we want to consider showing the values in this case, or is this fine as it is? @annebirzin 


<!-- Sent from Reviewable.io -->

lindseyjin reviewed (commented):
Sounds good, thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71012#-:-Mm3U_B92OKn9XQfSY30:b-k6rpao)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 3 stale) (waiting on @Annebirzin, @Azhng, @kevin-v-ngo, @maryliag, and @matthewtodd)



<!-- Sent from Reviewable.io -->

lindseyjin commented:
bors r+
